File: Stage25/Python/Problem1300.py
import sys
input = sys.stdin.readline

def main():
    N = int(input().strip())
    K = int(input().strip())

    left = 1
    right = N * N
    answer = 0

    while left <= right:
        mid = (left + right) // 2
        count = 0

        for i in range(1, N + 1):
            count += min(mid // i, N)

        if count >= K:
            answer = mid
            right = mid - 1
        else:
            left = mid + 1

    print(answer)

if __name__ == "__main__":
    main()

File: Stage25/C/Problem1300.c
#include <stdio.h>

int main(void) {
    int N, K;
    scanf("%d", &N);
    scanf("%d", &K);

    long long left = 1;
    long long right = (long long)N * N;
    long long answer = 0;

    while (left <= right) {
        long long mid = (left + right) / 2;
        long long count = 0;

        for (int i = 1; i <= N; i++) {
            long long tmp = mid / i;
            if (tmp > N) tmp = N;
            count += tmp;
        }

        if (count >= K) {
            answer = mid;
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }

    printf("%lld\n", answer);
    return 0;
}